test/integration/master/audit_test.go

```diff
@@ -164,62 +173,128 @@ rules:
+		auditPolicy := strings.Replace(auditPolicyPattern, "{version}", version, 1)
+		auditPolicy = strings.Replace(auditPolicy, "{level}", string(level), 1)
+		policyFile, err := ioutil.TempFile("", "audit-policy.yaml")
+		if err != nil {
+			t.Fatalf("Failed to create audit policy file: %v", err)
+		}
+		defer os.Remove(policyFile.Name())
+		if _, err := policyFile.Write([]byte(auditPolicy)); err != nil {
+			t.Fatalf("Failed to write audit policy file: %v", err)
+		}
+		if err := policyFile.Close(); err != nil {
+			t.Fatalf("Failed to close audit policy file: %v", err)
+		}
 
-	// perform configmap operations
-	configMapOperations(t, kubeclient)
+		// prepare audit log file
+		logFile, err := ioutil.TempFile("", "audit.log")
+		if err != nil {
+			t.Fatalf("Failed to create audit log file: %v", err)
+		}
+		defer os.Remove(logFile.Name())
 
-	// check for corresponding audit logs
-	stream, err := os.Open(logFile.Name())
-	if err != nil {
-		t.Fatalf("Unexpected error: %v", err)
+		// start api server
+		result := kubeapiservertesting.StartTestServerOrDie(t, nil,
+			[]string{
+				"--audit-policy-file", policyFile.Name(),
+				"--audit-log-version", version,
+				"--audit-log-mode", "blocking",
+				"--audit-log-path", logFile.Name()},
+			framework.SharedEtcd())
+		defer result.TearDownFn()
+
+		kubeclient, err := kubernetes.NewForConfig(result.ClientConfig)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		if enableMutatingWebhook {
+			if err := createV1beta1MutationWebhook(kubeclient, url+"/mutation"); err != nil {
+				t.Fatal(err)
+			}
+			// wait for webhook config to take effect
+			time.Sleep(5 * time.Second)
+		}
+
+		// perform configmap operations
+		configMapOperations(t, kubeclient)
+
+		// check for corresponding audit logs
+		stream, err := os.Open(logFile.Name())
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		defer stream.Close()
+		expected := prepareExpectedEvents(expectedEvents, level, enableMutatingWebhook)
+		missingReport, err := utils.CheckAuditLines(stream, expected, versions[version])
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		cleanupExpectedEvents(expected)
+		if len(missingReport.MissingEvents) > 0 {
+			t.Errorf(missingReport.String())
+		}
+		if closeFunc != nil {
+			closeFunc()
+		}
+	})
+}
+
+func prepareExpectedEvents(events []utils.AuditEvent, level auditinternal.Level, enableMutatingWebhook bool) []utils.AuditEvent {
+	if !enableMutatingWebhook {
+		return events
 	}
-	defer stream.Close()
-	missingReport, err := utils.CheckAuditLines(stream, expectedEvents, versions[version])
-	if err != nil {
-		t.Fatalf("Unexpected error: %v", err)
+	res := []utils.AuditEvent{}
+	// e is shadow copy of utils.AuditEvent from package-level variable expectedEvents
+	for _, e := range events {
+		if e.Verb != "create" && e.Verb != "update" {
+			continue
+		}
+		// TODO: the maps can be mutated by other test runs, because maps are copies of reference
```

Comment:
which maps?

Reply:
`WebhookMutationAnnotations` and `WebhookPatchAnnotations` in `test/utils.AuditEvent`